### ISSUE TITLE
Redirector depends on concrete Laravel\Lumen\Application which has no binding.

### ISSUE THREAD

**Issue by ifeltsweet** [laravel/lumen-framework#255]
`Laravel\Lumen\Http\Redirector` and `Laravel\Lumen\Routing\UrlGenerator` depend on `Laravel\Lumen\Application` to be injected in their constructors.

This breaks application if I in turn inject `Redirector` or `UrlGenerator` because Lumen never bound `Laravel\Lumen\Application` to container.

Lumen binds only the following classes to resolve application:

- `Illuminate\Container\Container`
- `Illuminate\Contracts\Container\Container`
- `Illuminate\Contracts\Foundation\Application`

So, to fix, we should either add `Laravel\Lumen\Application` to be another binding for application or make `Redirector` and `UrlGenerator` use one of the existing bindings.

**Comment by Stayallive** [comment]
To fix this I currently have this is my `bootstrap/app.php`.

```
$app = new Laravel\Lumen\Application(
    realpath(__DIR__ . '/../')
);

$app->singleton('Laravel\Lumen\Application', function() use ($app) {
    return $app;
});
```

**Comment by kevintanjung** [comment]
I had the same problem to when I tried to pass the `Laravel\Lumen\Http\Redirector` instance through a class constructor. It created another instance of `Laravel\Lumen\Application`, because the `Laravel\Lumen\Application` instance is bonded to `Illuminate\Contracts\Foundation\Application`. I also used @Stayallive solution to overcome this, because I rather not overriding the framework behavior.

I think either the attempt above should be done or the `Redirector` or `UrlGenerator` should have instead type-hinted to the Contract. I don't know which one is the better approach though. It does make sense to expect the `Laravel\Lumen\Application` for type-hinting, but it shouldn't make any problem either to let the Contract for type-hinting, because those classes will only be used in junction with Lumen right?
